recipe/meta.yaml

```diff
@@ -12,51 +15,82 @@ source:
-    # in the Python modules.
-    - arrow-cpp {{ version }}
-    - boost-cpp
-    - cython
-    - numpy 1.16.*
-    - python
-    - setuptools
-    - setuptools_scm
-    - six
+outputs:
+  - name: pyarrow
+    script: build-pyarrow.sh  # [not win]
+    script: bld.bat           # [win]
+    version: {{ version }}
+    build:
+      number: {{ number }}
+      string: "py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ number }}_{{ build_ext }}"
+      skip: true              # [win32]
+      skip: true              # [win and py<35]
+      skip: true              # [cuda_compiler_version not in (undefined, "None", "9.2")]
+      track_features:
+        {{ "- arrow-cuda" if cuda_enabled else "" }}
+      activate_in_script: True
 
-  run:
-    - {{ pin_compatible('numpy', lower_bound='1.16') }}
-    - boost-cpp
-    - futures  # [py==27]
-    - glog
-    - pandas
-    # empty parquet-cpp metapackage, force old versions to be uninstalled
-    - parquet-cpp 1.5.1.*
-    - python
-    - six
-    - snappy
+    requirements:
+      build:
+        - cmake
+        - ninja
+        - make                # [unix]
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
+        # pyarrow does not require nvcc but it needs to link against libraries in arrow-cpp=*=*cuda
+        - {{ compiler("cuda") }}  # [cuda_compiler_version is not undefined and cuda_compiler_version == '9.2']
+      host:
+        # directly pin boost-cpp as we also seem to directly include boost symbols
+        # in the Python modules.
+        - arrow-cpp {{ version }}
+        - arrow-cpp-proc * {{ build_ext }}
+        - boost-cpp
+        - cython
+        - numpy >=1.16
+        - python
+        - setuptools
+        - setuptools_scm
+        - six
+      run:
+        - {{ pin_compatible('numpy', lower_bound='1.16') }}
+        - boost-cpp
+        - arrow-cpp {{ version }}
+        - futures  # [py==27]
+        - glog
+        - pandas
+        # empty parquet-cpp metapackage, force old versions to be uninstalled
+        - parquet-cpp 1.5.1.*
+        - python
+        - six
+        - snappy
+      run_constrained:
+        - arrow-cpp-proc * {{ build_ext }}
 
-test:
-  imports:
-    - pyarrow
-    - pyarrow.dataset  # [not py==27]
-    - pyarrow.flight   # [not py==27]
-    - pyarrow.gandiva  # [(not py==27) and (not (aarch64 or ppc64le))]
-    - pyarrow.orc      # [unix]
-    - pyarrow.parquet
-    - pyarrow.plasma   # [unix]
-    - pyarrow.fs
-    - pyarrow._s3fs
-    - pyarrow._hdfs
+    test:
+      imports:
+        - pyarrow
+        - pyarrow.dataset  # [not py==27]
+        - pyarrow.flight   # [not py==27]
+        - pyarrow.gandiva  # [(not py==27) and (not (aarch64 or ppc64le))]
+        - pyarrow.orc      # [unix]
+        - pyarrow.parquet
+        - pyarrow.fs
+        - pyarrow._s3fs
+        - pyarrow._hdfs
+        # When using cuda, importing pyarrow.plasma and pyarrow.cuda
+        # packages require libcuda.so.1 that is not available in CI
```

Comment:
If the cuda compiler is required doesn't it symlink in `libcuda` to a path found during a typical search?

Reply:
Trying to import `pyarrow.cuda` from the recipe tests leads to 
```
ImportError: libcuda.so.1: cannot open shared object file: No such file or directory
```
The libcuda symlink exists for the build but not for the test section.

Adding nvcc and cudatoolkit to test requirements triggers the same error. Nor will the test work from the build script.

Is it expected that the "usage of CUDA support" (rather than the "availability of CUDA support") can be tested from CI instances?

Packages like arrow-cpp and omniscidb-cuda test only the availability of the CUDA support while importing "pyarrow.cuda" qualifies to testing the usage, IIUC.

Reply:
We should be able to at least find the `libcuda.so` shared library to prevent this error, though we shouldn't call `cuInit` as that would likely fail.

This would be from `nvcc`, not `cudatoolkit`.

Reply:
I verified that `${CONDA_BUILD_SYSROOT}/lib/libcuda.so` exists and it is a symbolic link to `${CUDA_HOME}/lib64/stubs/libcuda.so` as expected from `nvcc`.

`ldd` on the built `pyarrow._cuda` shared library shows `libcuda.so.1 => not found`.

I am not sure why `ldd` does not pickup `libcuda.so` from stubs.

Reply:
This could be a weird CMake thing where it links against `libcuda.so.1` instead of `libcuda.so` (we see the same thing for our RAPIDS libs), where we should maybe be symlinking `libcuda.so.1` and `libcuda.so.$DRIVER_VERSION` in the nvcc package.

Could you try symlinking in `libcuda.so.1` to see if that allows it to be found?

Reply:
> I am not sure why ldd does not pickup libcuda.so from stubs.

> This could be a weird CMake thing where it links against libcuda.so.1 instead of libcuda.so

Both incorrect. Have a look at SONAME. `libcuda.so` is for linking. It is needed only at downstream package build time. It has a soname of `libcuda.so.1` which was what the dependencies look for at load time. Most of the time `libfoo.so` is a symlink to `libfoo.so.<num>`, but that means you need to have `libfoo.so.<num>` when building and CUDA provides the stub libraries which are lightweight libraries for linking that can't be used at runtime. The ones for runtime are at `/usr/local/cuda-10.0/compat/`.

As to the problem at hand, this seems to be a problem with the `nvidia/cuda` docker images. `ld.so.conf.d` entries don't have the place where `libcuda.so.1` lives. Please open an issue in https://github.com/NVIDIA/nvidia-docker .

Reply:
PRs are also welcome to https://github.com/conda-forge/docker-images to fix this in our docker image.

Reply:
I think the issue is this is the driver library which is typically installed alongside the driver with kernel modules. I'm pretty sure using the container runtime with gpus typically maps this file in to make sure to correspond with the driver version installed on the host system.

Since these builds are running on CPU only nodes, presumably without the driver installed, we're almost definitely not using the gpu container runtime so there's no mapping of the driver and driver libs into the container. The container installs the toolkit which includes a stub for the driver library at `${CUDA_HOME}/lib64/stubs/libcuda.so` which we symlink to `${CONDA_BUILD_SYSROOT}/lib/libcuda.so`, but there's no non-stub library in general.

Reply:
How sure are you? I tried on a computer with a GPU and the driver versions are different (One on host and one on docker). (`libcuda.so.410.129` vs `libcuda.so.390.116`)

Reply:
I'm guessing the docker container actually had both and you ran a CUDA 9.2 image. Did a quick test with my host machine running driver version 440.64.00 and used docker image `nvidia/cuda:10.0-base-ubuntu16.04`. Dockerfile is here: https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/ubuntu16.04/10.0/base/Dockerfile.

Without passing `--gpus` option to docker, no `libcuda` files in the standard install location of `/usr/lib/x86_64-linux-gnu`.

With passing `--gpus` option to docker, the following files in the standard install location of `/usr/lib/x86_64-linux-gnu`:
```
libcuda.so
libcuda.so.1
libcuda.so.410.129
libcuda.so.440.64.00
```

This makes sense because `410.129` is the first driver version to support CUDA 10.0 and then `440.64.00` is my host driver version.

Reply:
The commit 52a2ede downloads cuda-compat-10-0_410.48-1_amd64.deb and after pointing LD_LIBRARY_PATH to `cuda-10.0/compat`, one can import `pyarrow.cuda` from the build script.
While not an ideal solution, it shows how to make `pyarrow.cuda` test to work.

Reply:
> @isuruf: PRs are also welcome to https://github.com/conda-forge/docker-images to fix this in our docker image.

I think I'm hitting the same topic in https://github.com/conda-forge/faiss-split-feedstock/pull/1, and I'm happy to open a PR, but I'm not sure yet what exactly needs to be done. Download drivers and install them?

Reply:
@h-vetinari, yes, downloading cuda-compat and install it is what needs to be done.

Another question is if it is going to acceptable to mix cuda 9.2 and 10.0 libraries? Although, the mix would resolve this issue (tested).

Reply:
> @pearu: @h-vetinari, yes, downloading cuda-compat and install it is what needs to be done.

This might be true for the 9.2 case, but I'm also struggling with missing drivers (I think) even on 10.x. Also, I haven't yet managed to wrap my head around this:
> @isuruf: As to the problem at hand, this seems to be a problem with the `nvidia/cuda` docker images. `ld.so.conf.d` entries don't have the place where `libcuda.so.1` lives.

Reply:
Yeah, the problem is that `libcuda.so.1` does not exist in the `nvidia/cuda` docker images, so there is nowhere to point from `ld.so.conf.d` ..

Reply:
The [`nvidia/cuda:9.2-devel-centos6`](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/centos6/9.2/devel/Dockerfile) image is based on `nvidia/cuda:9.2-runtime-centos6`, and this [includes](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/centos6/9.2/runtime/Dockerfile#L6) `cuda-libraries-$CUDA_PKG_VERSION.rpm`, at least at in the Dockerfile, but like @pearu said, this doesn't work correctly...

Starting a container with normal `docker run` (without the `--gpu` argument that @kkraus14 mentions, because `.scripts/run_docker_build.sh` doesn't do that either), we only get a stub:
```
[root@f2a51f1cbd4e /]# find /usr/ -name 'libcuda.so*'
/usr/local/cuda-9.2/targets/x86_64-linux/lib/stubs/libcuda.so
[root@56ddd2cd52b0 /]# ls -la /usr/local/cuda-9.2/targets/x86_64-linux/lib/stubs/libcuda.so
-rwxr-xr-x 1 root root 43192 Jun 13  2018 /usr/local/cuda-9.2/targets/x86_64-linux/lib/stubs/libcuda.so
```

As a comparison, within `nvidia/cuda:10.2-devel-centos6` we have
```
[root@7c2172c6db41 /]# find /usr/ -name 'libcuda.so*'
/usr/local/cuda-10.2/targets/x86_64-linux/lib/stubs/libcuda.so
/usr/local/cuda-10.2/compat/libcuda.so
/usr/local/cuda-10.2/compat/libcuda.so.1
/usr/local/cuda-10.2/compat/libcuda.so.440.33.01
[root@7c2172c6db41 /]# ls -la /usr/local/cuda-10.2/compat/libcuda*
lrwxrwxrwx 1 root root       20 Nov 27 07:29 /usr/local/cuda-10.2/compat/libcuda.so -> libcuda.so.440.33.01
lrwxrwxrwx 1 root root       20 Nov 27 07:29 /usr/local/cuda-10.2/compat/libcuda.so.1 -> libcuda.so.440.33.01
-rwxr-xr-x 1 root root 16591136 Nov 13  2019 /usr/local/cuda-10.2/compat/libcuda.so.440.33.01
```

Considering the (lack of) [activity](https://gitlab.com/nvidia/container-images/cuda/activity) upstream, I'll make a PR to https://github.com/conda-forge/docker-images directly.

Reply:
Opened a PR https://github.com/conda-forge/docker-images/pull/135.

Reply:
Now pending on https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/93

Reply:
Also pending https://github.com/conda-forge/docker-images/pull/136